timemachine/fe/rest/single_topology.py

```diff
@@ -0,0 +1,182 @@
+        return ring_bonds_c
+
+    @cached_property
+    def rotatable_bonds(self) -> set[CanonicalBond]:
+        rotatable_bonds_a = {bond.translate(self.a_to_c) for bond in get_rotatable_bonds(self.mol_a)}
+        rotatable_bonds_b = {bond.translate(self.b_to_c) for bond in get_rotatable_bonds(self.mol_b)}
+        rotatable_bonds_c = rotatable_bonds_a | rotatable_bonds_b
+        return rotatable_bonds_c
+
+    @cached_property
+    def propers(self) -> NDArray[np.int32]:
+        # TODO: refactor SingleTopology to compute src and dst alignment at initialization
+        return super().setup_intermediate_state(0.0).proper.potential.idxs
+
+    @cached_property
+    def target_proper_idxs(self) -> list[int]:
+        return [
+            idx
+            for idx, (_, j, k, _) in enumerate(self.propers)
+            for bond in [mkbond(j, k)]
+            if bond in self.rotatable_bonds or bond in self.aliphatic_ring_bonds
+        ]
+
+    @cached_property
+    def target_propers(self) -> NDArray[np.int32]:
+        return self.propers[self.target_proper_idxs, :]
+
+    def get_energy_scale_factor(self, lamb: float) -> float:
+        temperature_factor = self._temperature_scale_interpolation_fxn(lamb).item()
+        return 1.0 / temperature_factor
+
+    def setup_intermediate_state(self, lamb: float) -> VacuumSystem[NonbondedPairListPrecomputed, HarmonicAngleStable]:
+        ref_state = super().setup_intermediate_state(lamb)
+        energy_scale = self.get_energy_scale_factor(lamb)
+
+        assert ref_state.proper
+        proper = replace(
+            ref_state.proper,
+            params=jnp.asarray(ref_state.proper.params).at[self.target_proper_idxs, 0].mul(energy_scale),
+        )
+
+        nonbonded = replace(
+            ref_state.nonbonded,
+            params=jnp.asarray(ref_state.nonbonded.params)
+            .at[:, NBParamIdx.Q_IDX]
+            .mul(energy_scale)  # scale q_ij
+            .at[:, NBParamIdx.LJ_EPS_IDX]
+            .mul(energy_scale),  # scale eps_ij
+        )
+
+        return replace(ref_state, proper=proper, nonbonded=nonbonded)
+
+    def combine_with_host(
+        self,
+        host_system: VacuumSystem,
+        lamb: float,
+        num_water_atoms: int,
+        ff: Forcefield,
+        omm_topology: app.topology.Topology,
+    ) -> HostGuestSystem:
+        ref_state = super().combine_with_host(host_system, lamb, num_water_atoms, ff, omm_topology)
+        num_atoms_host = host_system.nonbonded.potential.num_atoms
+        energy_scale = self.get_energy_scale_factor(lamb)
+
+        nonbonded_host_guest_ixn = replace(
+            ref_state.nonbonded_host_guest_ixn,
+            params=jnp.asarray(ref_state.nonbonded_host_guest_ixn.params)
+            .at[num_atoms_host:, NBParamIdx.Q_IDX]
+            .mul(energy_scale)  # scale ligand charges
+            .at[num_atoms_host:, NBParamIdx.LJ_EPS_IDX]
+            .mul(energy_scale),  # scale ligand epsilons
```

Comment:
[style] instead of inlining the index expression `num_atoms_host:`, consider defining an index array `ligand_idxs` or similar and referencing it here. 

[just to confirm] To scale this interaction group potential by scaling the q and eps parameters, would all of the following three implementations be functionally equivalent? (1) scale the ligand's `q, eps` parameters by `energy_scale`, (2) scale the environment's `q, eps` parameters by `energy_scale`, (3) scale all particles' `q, eps` parameters by `sqrt(energy_scale)`?

Reply:
(If so, implementation option (3) seems potentially safer -- wouldn't be sensitive to possible errors in computing ligand idxs vs. host idxs.)

Reply:
> (If so, implementation option (3) seems potentially safer -- wouldn't be sensitive to possible errors in computing ligand idxs vs. host idxs.)

Ah, that's a good point. I had gone with (1) just to minimize the number of parameters changed, but I think your argument for (3) is stronger. Will fix.

Reply:
Changed to (3) in https://github.com/proteneer/timemachine/pull/1424/commits/4e07f8f261bc83784bfda3390b28c17f4623c79c.